cpu/rp2350_common/periph/uart.c

```diff
@@ -0,0 +1,243 @@
+    ctx[uart].arg = arg;
+
+    uart_init_pins(uart);
+
+    if (uart_mode(uart, UART_DATA_BITS_8, UART_PARITY_NONE, UART_STOP_BITS_1) != UART_OK) {
+        return UART_NOMODE;
+    }
+
+    /* enable RX and IRQs, if needed */
+    if (rx_cb != NULL) {
+        _irq_enable(uart);
+        /* clear any pending data and IRQ to avoid receiving a garbage char */
+        uint32_t status = dev->UARTRIS;
+        dev->UARTICR = status;
+        (void)dev->UARTDR;
+        atomic_set(&dev->UARTCR, UART_UARTCR_RXE_BITS);
+    }
+
+    return UART_OK;
+}
+
+void uart_write(uart_t uart, const uint8_t *data, size_t len) {
+    UART0_Type *dev = uart_config[uart].dev;
+    for (size_t i = 0; i < len; i++) {
+        dev->UARTDR = data[i];
+        /* Wait until the TX FIFO is empty before sending the next byte */
+        while (!(dev->UARTFR & UART_UARTFR_TXFE_BITS)) {
+        }
+    }
+}
+
+void uart_poweron(uart_t uart) {
+    assert((unsigned)uart < UART_NUMOF);
+    /* Get into a save state where we know whats up */
+    _reset_uart(uart);
+    UART0_Type *dev = uart_config[uart].dev;
+    /* Restore config from registers */
+    dev->UARTIBRD = uartibrd;
+    dev->UARTFBRD = uartfbrd;
+    dev->UARTLCR_H = uartlcr_h;
+    dev->UARTCR = uartcr;
+    /* restore IRQs, if needed */
+    if (ctx[uart].rx_cb != NULL) {
+        _irq_enable(uart);
+    }
+    uart_init_pins(uart);
+}
+
+void uart_deinit_pins(uart_t uart) {
+    assert((unsigned)uart < UART_NUMOF);
+    /* @TODO */
+    /* gpio_reset_all_config(uart_config[uart].tx_pin); */
+    SIO->GPIO_OE_CLR = 1LU << uart_config[uart].tx_pin;
+    if (ctx[uart].rx_cb) {
+        /* gpio_reset_all_config(uart_config[uart].rx_pin); */
+    }
+}
+
+void uart_poweroff(uart_t uart) {
+    assert((unsigned)uart < UART_NUMOF);
+    UART0_Type *dev = uart_config[uart].dev;
+    /* backup configuration registers */
+    uartibrd = dev->UARTIBRD;
+    uartfbrd = dev->UARTFBRD;
+    uartlcr_h = dev->UARTLCR_H;
+    uartcr = dev->UARTCR;
+    /* disconnect GPIOs and power off peripheral */
+    uart_deinit_pins(uart);
+    _reset_uart(uart);
+}
+
+void isr_handler(uint8_t num) {
+    UART0_Type *dev = uart_config[num].dev;
+
+    uint32_t status = dev->UARTMIS;
+    dev->UARTICR = status;
+
+    if (status & UART_UARTMIS_RXMIS_BITS) {
+        uint32_t data = dev->UARTDR;
+        // if (data & (UART0_UARTDR_BE_Msk | UART0_UARTDR_PE_Msk | UART0_UARTDR_FE_Msk)) {
+        //     DEBUG_PUTS("[rpx0xx] uart RX error (parity, break, or framing error");
+        // }
+        // else {
+        printf("UART%d received: %c\n", num, (char)(data & 0xFF));
+        ctx[num].rx_cb(ctx[num].arg, (uint8_t)data);
+        // }
+    }
+}
+
+/** Overwrites the WEAK_DEFAULT isr_uart0 */
+void isr_uart0(void) {
+    isr_handler(0);
+    #ifdef RP2350_USE_ARM
+    cortexm_isr_end();
+    #endif
+}
+
+void isr_uart1(void) {
+    isr_handler(1);
+    #ifdef RP2350_USE_ARM
+    cortexm_isr_end();
+    #endif
```

Comment:
```suggestion
#ifdef RP2350_USE_ARM
    cortexm_isr_end();
#endif
```

Considering that the code is shared, it would make sense to have this as a runtime check when using RISC-V and ARM at the same time (maybe. eventually.)